### PR TITLE
Protect POST requests routes

### PR DESCRIPTION
**Pull request by HuangStanley050**
Implemented a middleware for two lambdas, edit and adding hack ideas. Check the token issued by Cognito before allowing the two lambda to process.

### CODE REVIEW

File: checkAuth.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    statementOne.Resource = resource;
+    policyDocument.Statement[0] = statementOne;
+    authResponse.policyDocument = policyDocument;
+  }
+  return authResponse;
+};
+export const auth = async (event, context) => {
+  const token = event.authorizationToken;
+
+  if (!token) return "Not authorized, no token!!";
```
Comment (jpeyper) [inline checkAuth.js]: how does this display? it should be a "401 Unauthorized" response.

File: checkAuth.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  let result = await axios.get(
+    `https://cognito-idp.${process.env.AWS_REGION}.amazonaws.com/${process.env.AWS_USER_POOL_ID}/.well-known/jwks.json`
+  );
+  const pem = jwkToPem(result.data.keys[1]);
+  try {
+    const decoded = jwt.verify(tokenValue, pem, { algorithm: ["RS256"] });
+
+    return generatePolicy(decoded.sub, "Allow", event.methodArn);
+  } catch (err) {
+    return "token not valid";
```
Comment (jpeyper) [inline checkAuth.js]: how does this display? it should be a "401 Unauthorized" response.

File: checkAuth.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  if (!token) return "Not authorized, no token!!";
+  const tokenParts = event.authorizationToken.split(" ");
+  const tokenValue = tokenParts[1];
+
+  let result = await axios.get(
+    `https://cognito-idp.${process.env.AWS_REGION}.amazonaws.com/${process.env.AWS_USER_POOL_ID}/.well-known/jwks.json`
+  );
+  const pem = jwkToPem(result.data.keys[1]);
+  try {
+    const decoded = jwt.verify(tokenValue, pem, { algorithm: ["RS256"] });
```
Comment (jpeyper) [inline checkAuth.js:33]: what are we actually verifying here, that the token came from our cognito?

question: should anyone be able to edit a hack? or only the hacks original author? or a team mate?

File: serverless.yml [hunk 4]
```diff
@@ -67,6 +74,8 @@ functions:
           path: addhack
           method: post
           cors: true
+          authorizer: authorizerFunc
+
   joinhack:
```
Comment (jpeyper) [inline serverless.yml:79]: should this one be locked down as well?

**Review by jpeyper — APPROVED** [review]

File: checkAuth.js (hunk 1 above)
Comment (HuangStanley050) [reply checkAuth.js]:
```
{
    "statusCode": 403,
    "error": "Forbidden",
    "message": "No principalId set on the Response"
}
```

That's the error but I will do more research to see if we can pass in our custom error like "token not valid and with a 401 error

File: checkAuth.js (hunk 2 above)
Comment (HuangStanley050) [reply checkAuth.js]: same as above

File: checkAuth.js (hunk 3 above)
Comment (HuangStanley050) [reply checkAuth.js:33]: Yes we are verifying the token pass by Cognito which will be attached to our axios request.

At the moment we haven't 100% decided who can edit the hack, for the time being only the login user, I think Joseph have some logic in the front end to determine that if the hack belongs to a certain user. But then again after this token verification might have to further implement logic to guard against unauthorised edit

File: serverless.yml (hunk 4 above)
Comment (HuangStanley050) [reply serverless.yml:79]: Yes that' valid, if that's not guarded then we will get empty userId.

File: checkAuth.js (hunk 1 above)
Comment (HuangStanley050) [reply checkAuth.js]: just need to 

```
return new Error("Unauthorized");
```

**Review by HuangStanley050 — COMMENTED** [review]
will not merged for now until we worked out the deal with front end and how to store the jwt and guard against accidentail refresh or closing of the tab

**Comment by jpeyper** [comment]
> will not merged for now until we worked out the deal with front end and how to store the jwt and guard against accidentail refresh or closing of the tab


chuck it into `sessionStorage` or look at [redux-persist](https://github.com/rt2zz/redux-persist)

**Pull request merged**